crates/matrix-sdk-ui/src/timeline/controller/observable_items.rs

```diff
@@ -0,0 +1,1500 @@
+}
+
+impl ObservableItems {
+    /// Create an empty `ObservableItems`.
+    pub fn new() -> Self {
+        Self {
+            // Upstream default capacity is currently 16, which is making
+            // sliding-sync tests with 20 events lag. This should still be
+            // small enough.
+            items: ObservableVector::with_capacity(32),
+            all_remote_events: AllRemoteEvents::default(),
+        }
+    }
+
+    /// Get a reference to all remote events.
+    pub fn all_remote_events(&self) -> &AllRemoteEvents {
+        &self.all_remote_events
+    }
+
+    /// Check whether there is timeline items.
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
+    }
+
+    /// Subscribe to timeline item updates.
+    pub fn subscribe(&self) -> VectorSubscriber<Arc<TimelineItem>> {
+        self.items.subscribe()
+    }
+
+    /// Get a clone of all timeline items.
+    ///
+    /// Note that it doesn't clone `Self`, only the inner timeline items.
+    pub fn clone(&self) -> Vector<Arc<TimelineItem>> {
+        self.items.clone()
+    }
+
+    /// Start a new transaction to make multiple updates as one unit.
+    pub fn transaction(&mut self) -> ObservableItemsTransaction<'_> {
+        ObservableItemsTransaction {
+            items: self.items.transaction(),
+            all_remote_events: &mut self.all_remote_events,
+        }
+    }
+
+    /// Replace the timeline item at position `timeline_item_index` by
+    /// `timeline_item`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `timeline_item_index > total_number_of_timeline_items`.
+    pub fn replace(
+        &mut self,
+        timeline_item_index: usize,
+        timeline_item: Arc<TimelineItem>,
+    ) -> Arc<TimelineItem> {
+        self.items.set(timeline_item_index, timeline_item)
+    }
+
+    /// Get an iterator over all the entries in this `ObservableItems`.
+    pub fn entries(&mut self) -> ObservableItemsEntries<'_> {
+        ObservableItemsEntries(self.items.entries())
+    }
+
+    /// Call the given closure for every element in this `ObservableItems`,
+    /// with an entry struct that allows updating that element.
+    pub fn for_each<F>(&mut self, mut f: F)
+    where
+        F: FnMut(ObservableItemsEntry<'_>),
+    {
+        self.items.for_each(|entry| f(ObservableItemsEntry(entry)))
+    }
+}
+
+// It's fine to deref to an immutable reference to `Vector`.
+impl Deref for ObservableItems {
+    type Target = Vector<Arc<TimelineItem>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.items
+    }
+}
+
+/// An “iterator“ that yields entries into an `ObservableItems`.
+pub struct ObservableItemsEntries<'a>(ObservableVectorEntries<'a, Arc<TimelineItem>>);
+
+impl ObservableItemsEntries<'_> {
+    /// Advance this iterator, yielding an `ObservableItemsEntry` for the next
+    /// item in the timeline, or `None` if all items have been visited.
+    pub fn next(&mut self) -> Option<ObservableItemsEntry<'_>> {
```

Comment:
Wait, why isn't this an `Iterator` implementation?

Reply:
As the doc now says:

```rust
/// An iterator that yields entries into an `ObservableItems`.
///
/// It doesn't implement [`Iterator`] though because of a lifetime conflict: the
/// returned `Iterator::Item` could live longer than the `Iterator` itself.
/// Ideally, `Iterator::next` should take a `&'a mut self`, but this is not
/// possible.
```

I think that's the same reason why [`eyeball_im::ObservableVectorEntries`](https://docs.rs/eyeball-im/latest/eyeball_im/struct.ObservableVectorEntries.html).